cli/list/list.go

```diff
@@ -0,0 +1,154 @@
+			if binActive == "master" {
+				fmt.Printf("	%s\n", util.Bold(color.GreenString("master - active")))
+			} else {
+				fmt.Printf("	%s\n", color.YellowString("master"))
+			}
+
+		}
+		for _, bin := range binList {
+			if bin == binActive {
+				fmt.Printf("	%s\n", util.Bold(color.GreenString(bin+" - active")))
+			} else {
+				fmt.Printf("	%s\n", color.YellowString(bin))
+			}
+		}
+	}
+}
+
+// sortBinaries sorts versions of binaries.
+func sortBinaries(binList []string) {
+	sort.SliceStable(binList, func(i, j int) bool {
+		varLeftString := strings.Split(binList[i], ".")
+		varRightString := strings.Split(binList[j], ".")
+		var right []uint64
+		var left []uint64
+
+		for _, ch := range varLeftString {
+			i, _ := strconv.Atoi(ch)
+			left = append(left, uint64(i))
+		}
+
+		for _, ch := range varRightString {
+			i, _ := strconv.Atoi(ch)
+			right = append(right, uint64(i))
+		}
+
+		largestLen := util.Max(len(left), len(right))
+
+		for i := 0; i < largestLen; i++ {
+			var valLeft, valRight uint64 = 0, 0
+			if i < len(left) {
+				valLeft = left[i]
+			}
+
+			if i < len(right) {
+				valRight = right[i]
+			}
+
+			if valLeft != valRight {
+				return valLeft > valRight
+			}
+		}
+
+		return false
+	})
+}
+
+// parseBinaries seeks through fileList returning array of found versions of program,
+// active version of the program (if found) and bool (whether master version was found).
+func parseBinaries(fileList []fs.DirEntry, programName string,
+	binDir string) ([]string, string, bool, error) {
+	var err error
+	var binariesVersion []string
+	binActive := ""
+	binMaster := false
+
+	for _, f := range fileList {
+		if strings.HasPrefix(f.Name(), programName) && f.Name() != programName &&
+			f.Name() != programName+"_master" {
+			binVersion := strings.TrimPrefix(f.Name(), programName+"_")
+			binVersion = strings.TrimPrefix(binVersion, "v")
+			binariesVersion = append(binariesVersion, binVersion)
+		}
+
+		// files w/o version in name are symlinks to the active binaries.
+		if f.Name() == programName {
+			binActive, err = util.ResolveSymlink(filepath.Join(binDir, programName))
+			binActive = strings.TrimPrefix(filepath.Base(binActive), programName+"_")
+
+			if err != nil {
+				return nil, "", false, err
+			}
+		}
+
+		// master versions of binaries are processed separately from others.
+		if f.Name() == programName+"_master" {
+			binMaster = true
+		}
+
+	}
+	return binariesVersion, binActive, binMaster, err
+}
+
+// ListBinaries outputs installed versions of programs from bin_dir.
+func ListBinaries(cmdCtx *cmdcontext.CmdCtx, cliOpts *config.CliOpts) error {
+	var err error
+
+	binDir := cliOpts.App.BinDir
+	binDirFilesList, err := os.ReadDir(cliOpts.App.BinDir)
+	if err != nil {
+		return err
+	}
+
+	tarantoolBinaries, tarantoolActive, tarantoolMaster, err := parseBinaries(binDirFilesList,
+		"tarantool", binDir)
+	if err != nil {
+		return err
+	}
+
+	ttBinaries, ttActive, ttMaster, err := parseBinaries(binDirFilesList,
```

Comment:
There is a code duplication. There is no need in two slices, two sorts and two prints. Just use the loop by the binary names. Loop body will be the same for all binaries, because logic is the same.

Reply:
Changed